### PR TITLE
Blood cult: cleans up blood rites code, fixes a bug or two, & adds blackbox logging for spells, runes, and shades

### LINKED ISSUE #76562: blood rites unteaches itself if your blood charges happen to drop to zero and that's annoying
the point of the spell is that you can charge it up by absorbing blood, then spend that blood on various effects

but if any of those effects happen to reduce your blood charge level to exactly zero (or below), the spell unteaches itself (as a stun hand or the like would) and you have to relearn it

### PR DESCRIPTION
**Pull request by wesoda25**
## About The Pull Request

Cleans up blood rites code and fixes #76562 by porting @carlarctg implementation in #79460 (if you want to be credited in cl lmk). Also potentially fixes an oversight/undocumented bug, see my comment in the /obj/item/melee/blood_magic/manipulator/proc/heal_cultist proc.

Adds blackbox logging for rune activation and spell use for blood cult. Also adds logging for cult shades, encompassing creation, deconversion, being turned into a construct, death, and even suicide.

## Why It's Good For The Game

I'm particularly interested in the shade data, but figured it'd be nice to have numbers on spells and runes as well. I'm not experienced with blackbox logging and am unsure if my implementation was the best, would appreciate some feedback. There is also some intentional overlap with game/attack logs currently that I can remove if need be.

Blood rites were abomination, and look a little better now. Cheers.

## Changelog
:cl:
fix: blood rites - will no longer delete after running out of charges
fix: blood rites - now heal shades as well
fix: blood rites - now charge 1.65x for self healing as opposed to the ~2x they were before 
code: blood rites code hurts my eyes a bit less now
/:cl:

### CODE REVIEW

File: code/modules/antagonists/cult/blood_magic.dm [hunk 1]
```diff
@@ -672,89 +683,127 @@
...
+	if(overall_damage == 0)
+		to_chat(user,span_cult("That cultist doesn't require healing!"))
+		return FALSE
+
+	var/healing_ratio = uses / overall_damage
+	var/healing_cost = round(overall_damage)
+	if(human_bloodbag == user)
+		to_chat(user,span_cult("<b>Your blood healing is far less efficient when used on yourself!</b>"))
+		healing_ratio *= 0.35 // self-healing is 65% less effective
+		healing_cost *= 1.65  // self-healing is 65% more expensive
```
Comment (wesoda25) [inline code/modules/antagonists/cult/blood_magic.dm:773]: Above previously was: 
```
ratio *= 0.35 // Healing is half as effective if you can't perform a full heal
uses -= round(overall_damage) // Healing is 65% more "expensive" even if you can still perform the full heal
```
... which doesn't make sense. I disregarded the top comment and made the bottom one reality in my implementation. Formerly (as I understood it), healing ratio was 35% as effective on yourself and cost twice as much (overall damage was subtracted twice). 

I've maintained the 35% and changed the cost to be 65% more expensive.

File: code/modules/antagonists/cult/blood_magic.dm [hunk 2]
```diff
@@ -672,89 +683,127 @@
...
+	healing_ratio = -1 * clamp(healing_ratio, 0, 1)
+	uses = clamp(uses - healing_cost, 0, INFINITY)
```
Comment (wesoda25) [inline code/modules/antagonists/cult/blood_magic.dm:775]: Misuses of clamp here? Will look into later.

Comment (wesoda25) [reply code/modules/antagonists/cult/blood_magic.dm:775]: nvm its not me its them

**Comment by ATH1909** [comment]
holy fuck thank you

File: code/modules/antagonists/cult/blood_magic.dm (hunk 1 above)
Comment (wesoda25) [reply code/modules/antagonists/cult/blood_magic.dm:773]: Upon my review my math isn't right here, marking as draft until I have time to fix.

**Comment by jlsnow301** [comment]
pre-early return code was wild. thank you for your service. could you write some autodoc comments for the procs too?

**Comment by wesoda25** [comment]
> pre-early return code was wild. thank you for your service. could you write some autodoc comments for the procs too?

Honestly thought that was what I was doing. I'll revisit the guidelines for that and get the code looking better sometime later 👍

File: code/modules/antagonists/cult/blood_magic.dm [hunk 3]
```diff
@@ -671,171 +686,244 @@
...
+			return
+	..()
+
+/**
+ * handles blood rites usage on constructs
+ *
+ * will only return TRUE if some amount healing is done
+ */
+/obj/item/melee/blood_magic/manipulator/proc/heal_construct(atom/target, mob/living/carbon/human/user)
+	var/mob/living/basic/construct/construct_thing = target
```
Comment (wesoda25) [inline code/modules/antagonists/cult/blood_magic.dm:728]: Need to change this to accept shades

**Comment by wesoda25** [comment]
Tested and ready, assuming I didn't miss something. There will be some change in blood rites behavior after this: now works on shades, no longer deletes when it runs out of charges, now charge 1.65x for self healing as opposed to... like 2+ that it was doing before? Idk what the math was doing really but it didn't match with the comments (I think). I labeled all the above stuff as fixes even if maybe it isn't? Ultimately feels more logical to me but I am open to debate and willing to change any of that stuff, ultimately just wanna fade a GBP hit.

File: code/modules/antagonists/cult/blood_magic.dm [hunk 4]
```diff
@@ -671,91 +686,147 @@
...
-		if(istype(target, /obj/effect/decal/cleanable/blood) || istype(target, /obj/effect/decal/cleanable/trail_holder) || isturf(target))
-			blood_draw(target, user)
-		..()
+	if(!proximity)
+		return
+
+	if(isconstruct(target) && !heal_construct(target, user))
+		return
+	if(istype(target, /obj/effect/decal/cleanable/blood) || istype(target, /obj/effect/decal/cleanable/trail_holder) || isturf(target))
+		blood_draw(target, user)
```
[the remark this answers is not in the document]
Comment (wesoda25) [reply code/modules/antagonists/cult/blood_magic.dm:705]: I let this be the answer scared me

**Pull request merged**